### ISSUE TITLE
Add the possibility to edit sample track with AFP.

### ISSUE THREAD

**Issue by Gabrielxd195** [LMMS/lmms#4971]
Good. I think the title is very clear, the idea would be to combine both tools, Sample Track and Audio File Processor, giving us the possibility to edit the tco of the ST audio wave in real time through the AFP parameters. (I think I've seen that in fl studio).
![AFP en SP editado](https://user-images.githubusercontent.com/26127693/57578667-9bbb4500-745e-11e9-97d8-ec4155ff5a64.png)
But before telling me that: '' AFP is an instrument ''. I would like to give my humble opinion. We have several editors: Piano Roll, BB Editor, etc, but to work with audio we still do not have an editor dedicated to it. AFP is the closest thing to '' edit '' audio; but there are some problems like:

- **Play samples using Piano Roll or BB Editor.**
![Reproduccion mediante piano roll](https://user-images.githubusercontent.com/26127693/57578696-49c6ef00-745f-11e9-87e7-8c0cdfe540c8.png)
This is useful for small samples such as kick, snare, etc; while for large samples such as voices and songs it is an annoyance, because you have to guess the wavelength and create a note so long to play it whole.

- **- You can not play in any position. (similar to #2596).**
When the playhead plays the note it starts to play, but you can not skip or play it from any other position.

- **- The samples only last one second in BB Editor.**
![reproduccion mediante bb editor](https://user-images.githubusercontent.com/26127693/57578725-cf4a9f00-745f-11e9-9328-b40d0a06a511.png)
In addition to not having a visual reference of the wave of the sample, it only lasts a second playing and stops.

None of these problems have them Sample Track and although I have had many corrections, it still does not happen to be a simple player with a chain of effects. I have seen that they have outstanding improvements #1471, and that they are going to add the recording #4531. and this is great!!

Although this recording feature is very useful. Sometimes it is necessary to edit the recorded audio but then you would have to export the recorded audio and import it as shown in AFP and then edit it. This can affect the workflow.

Although AFP is an instrument and not an editor in itself, but it has basic characteristics that SP needs. then I ask myself: why not use the resources that are already there, and instead of developing new gui (buttons, menus, windows, etc) because it is not better to combine the characteristics of AFP with SP, if AFP is anyway a sample processor ? How long would it take to implement the even more basic features outlined in # 1471? I am not a developer and I know that this would take much longer than combining AFP and SP. this would not replace the current way of using samples, but would give us another option to edit the Sample Track using AFP. This provides us with features such as: Amplify, trim, revert, loop, add to the mixer, add effects, change the pitch, do zoon, add filter, lfo, envelope, etc.

**Comment by BaraMGB** [comment]
LMMS is missing a sample editor, that's true. But I doubt the AFP is the right tool for this job. In my opinion, LMMS needs a Sample Editor for Sample Tracks side by side with Piano Roll Editor for Instrument Tracks and Automation Editor for Automation Tracks. Perhaps someone could provide a mockup for such an editor. So we can develop such a tool.

**Comment by Gabrielxd195** [comment]
> LMMS is missing a sample editor, that's true. But I doubt the AFP is the right tool for this job. In my opinion, LMMS needs a Sample Editor for Sample Tracks 

I agree friend. to have an own editor for the edition of audio in LMMS would be super great !!!, but what time it would take developers to develop at least the basic and essential characteristics to have a more or less decent editor ?.

I would have to create interfaces, icons, etc. and even though I'm not a developer, I know it would take years to implement. while AFP although not an editor but at least has the essential features to edit the sample track. if we combine apf with sample track I think they would be more useful. perhaps you do not think it would be better to use resources that are already there, than to have to create others from the beginning?

**Comment by BaraMGB** [comment]
> but what time it would take developers to develop at least the basic and essential characteristics to have a more or less decent editor ?.

developing open source software isn't a competition of time. The quality of the software is what matters. 
>I would have to create interfaces, icons, etc. and even though I'm not a developer

Feel free to make it in Gimp or whatever you are comfortable with. The first step is to make a mockup. Then we can discuss a way to implement it. Look at the Automation editor which we can use as a blue print. Try to implement buttons we need for a sample editor.

**Comment by Gabrielxd195** [comment]
> Feel free to make it in Gimp or whatever you are comfortable with. The first step is to make a mockup. Then we can discuss a way to implement it. Look at the Automation editor which we can use as a blue print. Try to implement buttons we need for a sample editor.

That is, do you allow me to draw or design the model? I'm not a developer, but I have something in mind about how the sample editor would be. If I were a developer, I would definitely collaborate with you.

**Comment by BaraMGB** [comment]
Why not? You can use this thread for it. Just post your draft here. I'm looking forward to it.

**Comment by zonkmachine** [comment]
Wouldn't it be a good first step to allow opening a sample in an external editor?

**Comment by Gabrielxd195** [comment]
> Por qué no? Puede utilizar este hilo para ello. Solo publica tu borrador aquí. Estoy deseando que llegue.

I can do it and I would love it, although for now I can only send a demo with edited photos, I can not send any codes or interfaces made with qtcreator or anything like that. there are no problems, can I do it this way? if so, I'll edit a good time.

**Comment by Gabrielxd195** [comment]
> ¿No sería un buen primer paso permitir abrir una muestra en un editor externo?

It would be good, but for now do not mark it please, first let me create the model and upload it.

**Comment by Gabrielxd195** [comment]
> Why not? You can use this thread for it. Just post your draft here. I'm looking forward to it.

Good. I already finished the models. one is the characteristics that the song editor should have, and the other is from the '' Sample Editor ''.

![grabacion de audio sin marcas](https://user-images.githubusercontent.com/26127693/58210294-08cca700-7cb7-11e9-8550-0b675460e50a.png)

![grabacion de audio con marcas](https://user-images.githubusercontent.com/26127693/58210329-1eda6780-7cb7-11e9-8407-0773fa9921bc.png)

**1. Sample Editor button.**
As a sample editor is an editor, it is best to place it next to the other editors.

**2. Record button.**
Once the tracks are in recording mode, this button activates the recording and the tracks start recording all simultaneously.

**3. Rewind and Fast forward buttons.**
This is basic in many daw, with this you can play with the speed of the playhead.

**4. Track recording button.**
This puts the tracks in recording mode, (it is more practical to have a GUI button outside than an option in the menu).

**5. Led indicator. # 1360**

This would be the prelude to the sample editor, where you prepare the tracks to be recorded, this would lead to multitrack recording # 3394.


While the model of the sample editor would be this:
![Maqueta del Editor de audio de lmms sin marcas](https://user-images.githubusercontent.com/26127693/58211164-30bd0a00-7cb9-11e9-82e9-97b58c71a2d6.png)

![Maqueta del Editor de audio de lmms con marcas](https://user-images.githubusercontent.com/26127693/58211181-37e41800-7cb9-11e9-9453-2c9ef45d50d8.png)

**1. Sample Editor button.**

**2. Buttons: Play, Record, Play record, Stop.**
These buttons would have the same functionality as in the other editors.

**3. Rewind and Fast forward.**

**4. Selection and Draft.**
This gives us the possibility to select any part of the sample or its entirety, and draft to delete the selected

**5. Buttons Trim sample and Mute samples.**
These take them out of audacity, they have the function of trimming or silencing the selected part of the sample.

**6. Effects Reverse sample and Reverse phase.**
These buttons are similar to the automation editor, but in this case, flipping horizontally is the same as reversing the sample; while flipping horizontally would change the phase to the sample.

**7. Quantize.**

**8. Automation of effects.**
In this case it would not be for pitch bend, but to automate the effects of the lower panel.

**9. Pitch Correction through piano roll. #4661**
This would be the correction of manual tone using piano roll, it would be the closest thing to melodyne or autotune.

**10. Choose Input device.**
Like audacity, we would have the possibility to change devices in an instant.

**11. Effects Chain button.**
This button would open the chain of effects of the editor, to add more effects to the sample.

**12. Basic effects of the sample.**
These basic effects would be: volume, bread, stereo amlitude, etc. would give us the ease of using them without the need of any plugin.

**13. Sampling frequency.**
Here you choose the sampling frequency of the sample to be recorded or recorded.

Good. For me this would be the ideal editor for LMMS, I could not do more models, because it took me a lot of time to do this. (excuse me if you are bothered by the spelling mistakes and the words in Spanish, is that I do not speak English).

As you can see, take some audacity elements, you will see the wave is similar. The ideal is to have a fully functional editor with all its elements in view, with some basic effects that are used very commonly. I do not know what time it would take to do all this, in case you learn my idea, because LMMS needs a sample editor. please do not stop commenting, ask me anything. regards

**Comment by BaraMGB** [comment]
This looks really great. Thank you for your work. So we can use this as a basement when we start to develop such an editor.

**Comment by Gabrielxd195** [comment]
> This looks really great. Thank you for your work. So we can use this as a basement when we start to develop such an editor.

Thanks friend, I admire your effort and dedication to LMMS. but do not you think that this specific issue should not have its own thread? if you want we make another thread talking about this model. I started talking about afp, and ended up making a sample track model.
